src/net.cpp

```diff
@@ -717,6 +747,15 @@ void CNode::PushVersion()
     int64_t nTime = GetAdjustedTime();
     CAddress addrYou = (addr.IsRoutable() && !IsProxy(addr) ? addr : CAddress(CService("0.0.0.0",0)));
     CAddress addrMe = GetLocalAddress(&addr);
+    if (addrMe != addrSeenByPeer)
+    {
+        if(fDebug10) LogPrintf("updating ExternalIP");
+        Discover();
+        if (fUseUPnP)
+        {
+            MapPort();
+        }
+    }
```

Comment:
There's a security issue here. `CNode::PushVersion()` is called when another node connects. A network of malicious nodes could connect/disconnect quickly to force the victim to query the external IP address lookup service used by `Discover()` and/or the UPnP router repeatedly in quick succession by spoofing the local address in the message. Under enough pressure, this could exhaust the victim's system resources or trigger a blacklist from the external services. 

Please take a look at [commit `845c86d1` in Bitcoin](https://github.com/bitcoin/bitcoin/pull/5161/commits/845c86d128fb97d55d125e63653def38729bd2ed) that uses a different strategy to detect and re-advertise external IP address changes. It also removes the problematic dependency on the external service. We can probably adopt some of these ideas until we have time in the future to port all of the newer net processing code.

Reply:
dropped `Discover()`, so the problem with blacklisting is solved.
The addresses reported are now stored in a map like `map<IP_remote, IP_reportet>`.

If Mallory pushes tells us different addresses only the most recent one is stored, thus casts only one vote on our address.
(unless Mallory also changes its address)

So we will rely on the answers of our peers.